### PR TITLE
Replace PointerMap with potentially-sparse-array-based Map polyfill

### PR DESCRIPTION
**Pull request by stuartpb**
Alternative to gh-189

### CODE REVIEW

File: src/pointermap.js [hunk 1]
```diff
@@ -1,56 +1,47 @@
 /**
  * This module implements an map of pointer states
  */
-var USE_MAP = window.Map && window.Map.prototype.forEach;
-var POINTERS_FN = function() { return this.size; };
+var USE_MAP = window.Map && window.Map.prototype.forEach &&
+  !Object.getOwnPropertyDescriptor(Map.prototype,'size').value;
```
Comment (scottgonzalez) [inline src/pointermap.js]: What is this guarding against?

Comment (stuartpb) [inline src/pointermap.js]: https://bugzilla.mozilla.org/show_bug.cgi?id=807001

File: src/pointermap.js [hunk 2]
```diff
@@ -1,56 +1,47 @@
...
+  has: function(k) {
+    return this.array[k] !== undefined;
   },
-  delete: function(inId) {
-    var i = this.keys.indexOf(inId);
-    if (i > -1) {
-      this.keys.splice(i, 1);
-      this.values.splice(i, 1);
+  delete: function(k) {
+    if (!this.has(k)){
```
Comment (scottgonzalez) [inline src/pointermap.js]: Isn't this inverted?

File: src/pointermap.js [hunk 3]
```diff
@@ -1,56 +1,47 @@
...
-    this.values.length = 0;
+    this.array.length = 0;
+    this.size = 0;
   },
 
   // return value, key, map
   forEach: function(callback, thisArg) {
-    this.values.forEach(function(v, i) {
-      callback.call(thisArg, v, this.keys[i], this);
+    return this.array.forEach(function(v, k) {
```
Comment (scottgonzalez) [inline src/pointermap.js]: Why is this returning?

File: src/pointermap.js (hunk 1 above)
Comment (scottgonzalez) [inline src/pointermap.js]: That's silly, why would we care about something that was fixed almost three years ago? Remove this.

File: src/pointermap.js (hunk 3 above)
Comment (stuartpb) [inline src/pointermap.js]: [Tail position](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-tail-position-calls).

File: src/pointermap.js (hunk 2 above)
Comment (stuartpb) [inline src/pointermap.js]: You're right - will fix.

File: src/pointermap.js (hunk 3 above)
Comment (scottgonzalez) [inline src/pointermap.js]: Do you care to explain in plain English instead of spec prose?

**Comment by scottgonzalez** [comment]
@stuartpb Are you able to attend our meeting today? It's on IRC at noon ET (#jquery-meeting on freenode).

File: src/pointermap.js (hunk 3 above)
Comment (stuartpb) [inline src/pointermap.js]: Ending a function in tail position allows for [tail call optimization](http://raganwald.com/2015/02/07/tail-calls-defult-arguments-recycling.html#tail-call-optimization) - basically, functions that end with "return" and a call to another function relinquish their stack space to the next function.

**Comment by stuartpb** [comment]
@scottgonzalez Sure, I can stay up for a few more hours.

**Comment by scottgonzalez** [comment]
Is there a reason not to just use an object instead of an array?

File: src/pointermap.js (hunk 1 above)
Comment (stuartpb) [inline src/pointermap.js]: > That's silly, why would we care about something that was fixed almost three years ago?

Why check `window.Map.prototype.forEach` then?

File: src/pointermap.js (hunk 3 above)
Comment (jdalton) [inline src/pointermap.js]: As far as I know no JS engine supports that optimization.

Comment (stuartpb) [inline src/pointermap.js]: No engine supports that optimization *yet*, but it's slated, so why not be ready for it?

Comment (jdalton) [inline src/pointermap.js]: > so why not be ready for it?

It may avoid call stack limits but there's no guarantee it'll bring a perf win.

Comment (narqo) [inline src/pointermap.js]: Is there any chance that there will be support of tail recursion in browsers, which don't support native `Map`? Or am I missing something?
